plugins/inputs/mongodb/mongodb.go

```diff
@@ -115,60 +115,73 @@ func (m *MongoDB) getMongoServer(url *url.URL) *Server {
 
 func (m *MongoDB) gatherServer(server *Server, acc telegraf.Accumulator) error {
 	if server.Session == nil {
-		var dialAddrs []string
-		if server.Url.User != nil {
-			dialAddrs = []string{server.Url.String()}
-		} else {
-			dialAddrs = []string{server.Url.Host}
+		dialInfo, err := m.parseConnectionString(server)
+		if err != nil {
+			return err
 		}
-		dialInfo, err := mgo.ParseURL(dialAddrs[0])
+
+		sess, err := mgo.DialWithInfo(dialInfo)
 		if err != nil {
-			return fmt.Errorf("Unable to parse URL (%s), %s\n",
-				dialAddrs[0], err.Error())
+			return fmt.Errorf("Unable to connect to MongoDB, %s\n", err.Error())
 		}
-		dialInfo.Direct = true
-		dialInfo.Timeout = 5 * time.Second
-
-		var tlsConfig *tls.Config
-
-		if m.Ssl.Enabled {
-			// Deprecated SSL config
-			tlsConfig = &tls.Config{}
-			if len(m.Ssl.CaCerts) > 0 {
-				roots := x509.NewCertPool()
-				for _, caCert := range m.Ssl.CaCerts {
-					ok := roots.AppendCertsFromPEM([]byte(caCert))
-					if !ok {
-						return fmt.Errorf("failed to parse root certificate")
-					}
+		server.Session = sess
+	}
+	return server.gatherData(acc, m.GatherPerdbStats)
+}
+
+func (m *MongoDB) parseConnectionString(server *Server) (*mgo.DialInfo, error) {
+	if server.Url == nil {
+		return nil, fmt.Errorf("Unable to connect to MongoDB because connection string was not set.")
+	}
+	var dialAddrs []string
+	if server.Url.User != nil {
+		dialAddrs = []string{server.Url.String()}
+	} else {
+		dialAddrs = []string{server.Url.Host}
```

Comment:
This seems wrong to me, because it will throw away other parts of the URL.  What was the cause of the authentication database not being parsed?